libs/datalayout/src/main/java/com/linkedin/openhouse/datalayout/generator/FileEntropyDataLayoutStrategyGenerator.java

```diff
@@ -0,0 +1,67 @@
+package com.linkedin.openhouse.datalayout.generator;
+
+import com.linkedin.openhouse.datalayout.datasource.FileStat;
+import com.linkedin.openhouse.datalayout.datasource.TableFileStats;
+import com.linkedin.openhouse.datalayout.datasource.TablePartitionStats;
+import com.linkedin.openhouse.datalayout.strategy.DataLayoutStrategy;
+import com.linkedin.openhouse.datalayout.util.DataCompactionConfigUtil;
+import com.linkedin.openhouse.datalayout.util.TableFileStatsUtil;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import lombok.Builder;
+import org.apache.spark.api.java.function.MapFunction;
+import org.apache.spark.sql.Dataset;
+import org.apache.spark.sql.Encoders;
+
+/**
+ * Data layout optimization strategies generator for OpenHouse. Generates a list of strategies with
+ * score.
+ */
+@Builder
+public class FileEntropyDataLayoutStrategyGenerator implements DataLayoutStrategyGenerator {
+  private static final long MB = 1024 * 1024;
+  private static final long FILE_BLOCK_SIZE_BYTES = 256 * MB;
+  private static final long FILE_BLOCK_MARGIN_BYTES = 10 * MB;
+  private static final long TARGET_BYTES_SIZE = 2 * FILE_BLOCK_SIZE_BYTES - FILE_BLOCK_MARGIN_BYTES;
+  private final TableFileStats tableFileStats;
+  private final TablePartitionStats tablePartitionStats;
+
+  /**
+   * Generate a list of data layout optimization strategies based on the table file stats and
+   * historic query patterns.
+   */
+  @Override
+  public List<DataLayoutStrategy> generate() {
+    return Collections.singletonList(generateCompactionStrategy());
+  }
+
+  /**
+   * Computes the score for a partitioned table as file entropy, the score is the mean-squared
+   * error. The higher the MSE, the worse the table state.
+   */
+  private DataLayoutStrategy generateCompactionStrategy() {
+
+    // Retrieve file sizes of all data files.
+    Dataset<Long> fileSizes =
+        tableFileStats.get().map((MapFunction<FileStat, Long>) FileStat::getSize, Encoders.LONG());
+
+    // Aggregate MSE values.
+    Double mse = 0.0;
+    Iterator<Long> itr = fileSizes.toLocalIterator();
+    while (itr.hasNext()) {
+      mse += Math.pow(TARGET_BYTES_SIZE - itr.next(), 2);
+    }
+    // Normalize.
+    mse = mse / fileSizes.count();
+
+    return DataLayoutStrategy.builder()
+        .config(
+            DataCompactionConfigUtil.createDefaultCompactionConfig(
+                TARGET_BYTES_SIZE,
+                tablePartitionStats.get().count(),
+                TableFileStatsUtil.getAggregatedFileStats(tableFileStats, TARGET_BYTES_SIZE)._1))
+        .score(mse)
```

Comment:
How is entroy used in the multi objective optimization where we need to rank based on compute cost objective and then pick top-K?
How do we compute cost and gain for the entropy strategy?

Reply:
Entropy is a version of gain, i.e., the higher the entropy, the higher the potential gain (a higher entropy means more deviation from the target file size). If you're calculating the score as gain/cost, you could also alternatively use entropy/cost for the top-k calculation and a higher score means that the table has higher priority for compaction according to its entropy.

Reply:
Sounds good so I would sort by entropy for ranking, and use the sum(compute cost) < compute budget for top-k selection. That works!

Impact in terms of file count reduction can be continued to be measured as sum(gains).